src/filters/matches.rs

```diff
@@ -0,0 +1,439 @@
+    type Error = eyre::Report;
+
+    fn try_from(value: proto::matches::DirectionalConfig) -> Result<Self, Self::Error> {
+        Ok(Self {
+            metadata_key: value.metadata_key.ok_or_else(|| {
+                ConvertProtoConfigError::new("Missing", Some("metadata_key".into()))
+            })?,
+            branches: value
+                .branches
+                .into_iter()
+                .map(proto::matches::Branch::try_into)
+                .collect::<Result<_, _>>()?,
+            fallthrough: value
+                .fallthrough
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("fallthrough".into())))?
+                .try_into()?,
+        })
+    }
+}
+
+/// Configuration for a specific direction
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+pub struct DirectionalConfig {
+    #[serde(rename = "metadataKey")]
+    pub metadata_key: String,
+    pub branches: Vec<Branch>,
+    #[serde(default)]
+    pub fallthrough: Fallthrough,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+pub struct Branch {
+    pub value: crate::metadata::Value,
+    pub filter: String,
+    pub config: Option<ConfigType>,
+}
+
+impl TryFrom<proto::matches::Branch> for Branch {
+    type Error = eyre::Report;
+
+    fn try_from(branch: proto::matches::Branch) -> Result<Self, Self::Error> {
+        Ok(Self {
+            value: branch
+                .value
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("value".into())))?
+                .try_into()?,
+            filter: branch
+                .filter
+                .ok_or_else(|| ConvertProtoConfigError::new("Missing", Some("filter".into())))?,
+            config: branch.config.map(ConfigType::Dynamic),
+        })
+    }
+}
+
+///  How the [`Matches`] filter should handle no branch being matched.
+#[derive(Debug, PartialEq)]
+pub enum Fallthrough {
+    /// The packet will be passed onto the next filter.
+    Pass,
+    /// The packet will be dropped. **Default behaviour**
+    Drop,
+    /// The filter specified in `filter` will be called.
+    Filter {
+        filter: String,
+        config: Option<ConfigType>,
+    },
+}
+
+impl Default for Fallthrough {
+    fn default() -> Self {
+        Self::Drop
+    }
+}
+
+impl TryFrom<proto::matches::directional_config::Fallthrough> for Fallthrough {
+    type Error = eyre::Report;
+
+    fn try_from(
+        branch: proto::matches::directional_config::Fallthrough,
+    ) -> Result<Self, Self::Error> {
+        use proto::matches::directional_config::Fallthrough as ProtoFallthrough;
+
+        Ok(match branch {
+            ProtoFallthrough::Pass(_) => Self::Pass,
+            ProtoFallthrough::Drop(_) => Self::Drop,
+            ProtoFallthrough::Filter(filter) => Self::Filter {
+                filter: filter
+                    .filter
+                    .ok_or_else(|| eyre::eyre!("Unexpected missing `filter` field"))?,
```

Comment:
```suggestion
                    .ok_or_else(|| eyre::eyre!("missing `filter` field in Fallthrough configuration"))?,
```
or something similar that mentions the context in which this happened